packages/ingest-mongodb-public/src/sources/MongoDbDotComDataSource.ts

```diff
@@ -0,0 +1,221 @@
+
+function makePageUrl(baseUrl: string, urlSlug: string): string {
+  const cleanedUrlSlug = urlSlug.startsWith("/") ? urlSlug.slice(1) : urlSlug;
+  const cleanedBaseUrl = baseUrl.endsWith("/") ? baseUrl.slice(0, -1) : baseUrl;
+  return `${cleanedBaseUrl}/${cleanedUrlSlug}`;
+}
+function makePageTitle(input: CustomerPage): string | undefined {
+  // easy case
+  if (input.title) {
+    return input.title;
+  }
+  // recurse!
+  if (input.components) {
+    return findPageTitleInTree(
+      input.components as unknown as DotComCmsComponent
+    );
+  }
+}
+function findPageTitleInTree(tree: DotComCmsComponent): string | undefined {
+  for (const key in tree) {
+    const value = tree[key];
+    if (key === "h1" && typeof value === "string") {
+      return value;
+    }
+    if (typeof value === "object") {
+      const foundTitle = findPageTitleInTree(value as DotComCmsComponent);
+      if (foundTitle) {
+        return foundTitle;
+      }
+    }
+  }
+}
+
+function makePageBody(input: CustomerPage): string {
+  if (!(input.body ?? input.components)) {
+    return "🤷";
+  }
+  assert(
+    input.body ?? input.components,
+    "input must have `body` or `components`"
+  );
+  // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
+  const pageBody = input.body
+    ? input.body
+    : // TODO: see if i can simplify w/o the map..i think can do parseCustomerPageRecursiveHelper(input.components)
+      // and this would work
+      input.components
+        ?.map((component) => {
+          return parseCustomerPageRecursiveHelper(component);
+        })
+        .join("\n\n")
+        .replaceAll(/\n{2,}/g, "\n\n")
+        .trim();
+  if (!pageBody) {
+    return "🤷";
+  }
+  assert(pageBody, "pageBody must be defined per above logic");
+  return removeMarkdownImagesAndLinks(striptags(pageBody));
+}
+
+function removeMarkdownImagesAndLinks(content: string): string {
+  const mdLink = /!?\[([\s\S]*?)\]\([\s\S]*?\)/g;
+
+  let cleanedContent = content.replaceAll(mdLink, (match, text) => {
+    // remove images
+    if (match.startsWith("!")) {
+      return "";
+    } else return text;
+  });
+  // remove unnecessary new lines
+  cleanedContent = cleanedContent.replaceAll(/\n{3,}/g, "\n\n");
+
+  return cleanedContent;
+}
+
+function parseCustomerPageRecursiveHelper(
+  tree: DotComCmsComponent,
+  result = ""
+): string {
+  // do not process Nav or Footer components
+  if (tree.key === "Nav" || tree.key === "Footer") {
+    return "";
+  }
+
+  if (tree.h1) {
+    result += `# ${tree.h1}\n\n`;
+  }
+  if (tree.h2) {
+    result += `## ${tree.h2}\n\n`;
+  }
+  if (tree.h3) {
+    result += `### ${tree.h3}\n\n`;
+  }
+  if (tree.h4) {
+    result += `#### ${tree.h4}\n\n`;
+  }
+  if (tree.h5) {
+    result += `##### ${tree.h5}\n\n`;
+  }
+  if (tree.h6) {
+    result += `###### ${tree.h6}\n\n`;
+  }
+  if (tree.title) {
+    result += `${tree.title}\n\n`;
+  }
+  if (tree.content) {
+    result += `${tree.content}\n\n`;
+  }
```

Comment:
Yar... this `if` chain upsets me a bit?

Might be too hacker, but you can make this extensible (and copy-paste-error-proof) with something like:

```ts
const prefixTable = {
  h1: "# ",
  h2: "## ",
  h3: "### ",
  h4: "#### ",
  h5: "##### ",
  h6: "###### ",
  title: "",
  content: "",
};

Object.entries(tree).forEach(([key, value]) => {
  if (typeof value === "object") {
    result += parseCustomerPageRecursiveHelper(value as DotComCmsComponent);
  } else if (typeof value === "string" && prefixTable[key]) {
    result += `${prefixTable[key]}${value}\n\n`;
  }
});
```